rpc/gnmi/gnmi-config-subscriptions.md

```diff
@@ -0,0 +1,187 @@
+message ConfigSubscriptionSyncDone {
+  // ID of a commit confirm operation as assigned by the client
+  // see Commit Confirm extension for more details.
+  string commit_confirm_id = 1;
+  // ID of a commit as might be assigned by the server
+  // when registering a commit operation.
+  string server_commit_id = 2;
+}
+```
+
+## 2.2 Actions
+
+### 2.2.1 ConfigSubscriptionStart
+
+A `ConfigSubscriptionStart` message is used by a gNMI client in a
+`SubscribeRequest` to indicate that it wants to start a ConfigSubscription.
+The target must respond exclusively with configuration data relevant to the
+created subscription.
+The target must respond with an initial set of updates followed by a
+`SubscribeResponse` where the `sync_update` field is set to `true`.
+However, if the `updates_only` field in the `SubscribeRequest` was set to
+`true`, the target should skip sending the initial updates and only send
+a `SubscribeResponse` with `sync_update` set to `true` as per the base gNMI
+specification.
+
+### 2.2.2 ConfigSubscriptionSyncDone
+
+A `ConfigSubscriptionSyncDone` message is used by a gNMI target in a
+`SubscribeResponse` to indicate a commit boundary to the client.
+A commit boundary marks the completion point of a specific set of
+configuration changes.
+It indicates that all changes within that set have been committed and
+that all notifications triggered by the commit have been streamed to
+the client.
+
+The `ConfigSubscriptionSyncDone` message includes two optional fields:
+
+* `commit_confirm_id`: A commit confirm ID assigned by the client which
+initiated the commit.
+The commit can be initiated via gNMI (using the CommitConfirmed Extension),
+NETCONF, or any other management interface. Applicable only if the commit
+confirmed option is used.
+* `server_commit_id`: An optional internal ID assigned by the target.
+
+In the case a commit happens before the `sync_response: true` and the server
+may include the committed changes together with the initial updates but has
+to send the related `ConfigSubscriptionSyncDone` after sending the
+`sync_response: true` response.
+
+## 3. Configuration changes scenarios
+
+### 3.1 Configuration changes without Commit Confirmed
+
+1) The client subscribes to path P1 with the `ConfigSubscription` extension
+present with the action `ConfigSubscriptionStart`.
+2) The server processes the subscription request as usual but will only send
+updates for the configuration schema nodes under the path P1.
+3) The client sends a Set RPC with the configuration changes to the path P1
+**without** the `CommitConfirm` extension.
+4) The server processes the Set RPC as usual and sends the updates for the
+configuration schema nodes under the path P1.
+5) After all the configuration updates are sent, the server sends the
+`ConfigSubscriptionSyncDone` message to the client in a SubscribeResponse
+message.
+This message does not contain a `commit_confirmed_id` and may contain a
+`server_commit_id`
```

Comment:
Maybe we should just have a "done" bool in this message as well as these IDs -- otherwise we have to do a check for a non-nil message, right?

Reply:
`3.1 Configuration changes without Commit Confirmed` is the case the change is made without CommitConfirmed and the device does not provide a `server_commit_id`. 
Checking a nil msg is equivalent to checking for a boolean. In both cases, the client has to check for a nil message.

Reply:
I don't think that's necessarily true -- if I can do `foo.GetMessage().GetBool()` then I don't have to do a nil-check for the `message` field in the Go API at least. However, regardless, my view is that having explicit fields makes debugging much easier, since seeing `done: true` is much clearer than having to remember that an empty message in a `oneof` means something :-)

Reply:
agree, will update the proto